crates/json-abi/src/state_mutability.rs

```diff
@@ -0,0 +1,183 @@
+use serde::{Deserialize, Serialize};
+
+const COMPAT_ERROR: &str = "state mutability cannot be both `payable` and `constant`";
+
+/// A JSON ABI function's state mutability.
+///
+/// This will serialize/deserialize as the `stateMutability` JSON ABI field's value, see
+/// [`as_json_str`](Self::as_json_str).
+/// For backwards compatible deserialization, see [`serde_state_mutability_compat`].
+#[derive(
+    Clone, Copy, Debug, Default, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize,
+)]
+#[serde(rename_all = "lowercase")]
+pub enum StateMutability {
+    /// Pure functions promise not to read from or modify the state.
+    Pure,
+    /// View functions promise not to modify the state.
+    View,
+    /// Nonpayable functions promise not to receive Ether.
+    ///
+    /// This is the solidity default: <https://docs.soliditylang.org/en/latest/abi-spec.html#json>
+    ///
+    /// The state mutability nonpayable is reflected in Solidity by not specifying a state
+    /// mutability modifier at all.
+    #[default]
+    NonPayable,
+    /// Payable functions make no promises.
+    Payable,
+}
+
+impl StateMutability {
+    /// Returns the string representation of the state mutability.
+    #[inline]
+    pub const fn as_str(self) -> Option<&'static str> {
+        if let Self::NonPayable = self {
+            None
+        } else {
+            Some(self.as_json_str())
+        }
+    }
+
+    /// Returns the string representation of the state mutability when serialized to JSON.
+    #[inline]
+    pub const fn as_json_str(self) -> &'static str {
+        match self {
+            Self::Pure => "pure",
+            Self::View => "view",
+            Self::NonPayable => "nonpayable",
+            Self::Payable => "payable",
+        }
+    }
+}
+
+/// [`serde`] implementation for [`StateMutability`] for backwards compatibility with older
+/// versions of the JSON ABI.
+///
+/// In particular, this will deserialize the `stateMutability` field if it is present,
+/// and otherwise fall back to the deprecated `constant` and `payable` fields.
+///
+/// Since it must be used in combination with `#[serde(flatten)]`, a `serialize` implementation
+/// is also provided, which will always serialize the `stateMutability` field.
+///
+/// # Examples
+///
+/// Usage: `#[serde(default, flatten, with = "serde_state_mutability_compat")]` on a
+/// [`StateMutability`] struct field.
```

Comment:
TIL this is possible

Reply:
I was praying this would work, it did :pray: